app/services/authentication/providers.rb

```diff
@@ -1,11 +1,70 @@
+# We require all authentication modules to make sure providers
+# are correctly preloaded both in development and in production and
+# ready to be used when needed at runtime
+Dir[Rails.root.join("app/services/authentication/**/*.rb")].each do |f|
+  require_dependency(f)
+end
+
 module Authentication
   module Providers
-    # TODO: [thepracticaldev/oss] raise exception if provider is available but not enabled for this app
-    # TODO: [thepracticaldev/oss] add available providers, enabled providers
+    # Retrieves a provider that is both available and enabled
     def self.get!(provider_name)
-      "Authentication::Providers::#{provider_name.titleize}".constantize
-    rescue NameError => e
-      raise ::Authentication::Errors::ProviderNotFound, e
+      name = provider_name.to_s.titleize
+
+      unless Authentication::Providers.const_defined?(name)
+        raise(
+          ::Authentication::Errors::ProviderNotFound,
```

Comment:
Is the view going to handle this or will we end up seeing this in Honeybadger? 

It looks like we would see it in Honeybadger which I think could get noisy if someone comes in and tries to throw junk into the form.

Reply:
The chances for this happen are really slim, the available providers are loaded on disk, someone should basically hijack the login URL to trigger this I think, but if it happens, it'll end up in DD first:

https://github.com/thepracticaldev/dev.to/blob/2d26318cf96c0f1c5c2e827b74bbfa6d27d292d3/app/services/authentication/authenticator.rb#L69

and then in the log file as we haven't decided what to do with the `OmniauthCallbacksController` yet:

https://github.com/thepracticaldev/dev.to/blob/2d26318cf96c0f1c5c2e827b74bbfa6d27d292d3/app/controllers/omniauth_callbacks_controller.rb#L81-L86